### PR TITLE
SONARJS-524 Exception during building string for array literal with empty elements

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ArrayLiteralTreeImpl.java [hunk 1]
```diff
@@ -33,24 +34,34 @@
...
 
-  public ArrayLiteralTreeImpl(List<ExpressionTree> elements, List<InternalSyntaxToken> commas) {
-    this.elements = new SeparatedList<>(elements, commas);
+  public ArrayLiteralTreeImpl(List<Tree> elementsAndCommas) {
+    this.elementsAndCommas = elementsAndCommas;
+    this.elements = getExpressionTrees(elementsAndCommas);
+  }
+
+  private List<ExpressionTree> getExpressionTrees(List<Tree> elementsAndCommas) {
+    List<ExpressionTree> elements = new ArrayList<>();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ArrayLiteralTreeImpl.java]: MAJOR Rename "elements" which hides the field declared at line 40. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 2]
```diff
@@ -583,15 +577,13 @@ public ArrayLiteralTreeImpl completeArrayLiteral(InternalSyntaxToken openBracket
...
-    List<InternalSyntaxToken> commas = Lists.newArrayList();
+    List<Tree> commas = Lists.newArrayList();
 
     for (InternalSyntaxToken comma : commaTokens) {
-      elements.add(new UndefinedTreeImpl());
       commas.add(comma);
     }
 
-    return new ArrayLiteralTreeImpl(elements, commas);
+    return new ArrayLiteralTreeImpl(commas);
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @vilchik-elena can't we directly use commaTokens?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @pynicolas indeed:)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 3]
```diff
@@ -524,35 +522,32 @@ public ArrayLiteralTreeImpl newArrayLiteralWithElements(Optional<List<InternalSy
...
 
         // First comma
-        commas.add(t.first().get(0));
+        elementsAndCommas.add(t.first().get(0));
         t.first().remove(0);
 
         for (InternalSyntaxToken comma : t.first()) {
-          elements.add(new UndefinedTreeImpl());
-          commas.add(comma);
+          elementsAndCommas.add(comma);
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @vilchik-elena why not `elementsAndCommas.addAll(t.first())`?

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 4]
```diff
@@ -561,18 +556,17 @@ public ArrayLiteralTreeImpl newArrayLiteralWithElements(Optional<List<InternalSy
...
       // Trailing comma after the last element
-      commas.add(restCommas.get().get(0));
+      elementsAndCommas.add(restCommas.get().get(0));
 
       // Elided array element at the end
       if (nbEndingComma > 1) {
         for (InternalSyntaxToken comma : restCommas.get().subList(1, nbEndingComma)) {
-          elements.add(new UndefinedTreeImpl());
-          commas.add(comma);
+          elementsAndCommas.add(comma);
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @vilchik-elena why not `elementsAndCommas.addAll(...)`?

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ArrayLiteralTreeImpl.java [hunk 5]
```diff
@@ -33,24 +34,34 @@
...
+    this.elements = Collections.emptyList();
+    this.elementsAndCommas = Collections.emptyList();
     this.closeBracket = closeBracket;
   }
 
-  public ArrayLiteralTreeImpl(List<ExpressionTree> elements, List<InternalSyntaxToken> commas) {
-    this.elements = new SeparatedList<>(elements, commas);
+  public ArrayLiteralTreeImpl(List<Tree> elementsAndCommas) {
+    this.elementsAndCommas = elementsAndCommas;
+    this.elements = getExpressionTrees(elementsAndCommas);
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ArrayLiteralTreeImpl.java]: @vilchik-elena another possible approach: 
`ImmutableList.copyOf(Iterables.filter(elementsAndCommas, ExpressionTree.class))`

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ArrayLiteralTreeImpl.java]: @pynicolas thanks!

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java (hunk 4 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @pynicolas thanks!

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @pynicolas thanks!

**Pull request merged**